libevmasm/Assembly.cpp

```diff
@@ -298,6 +502,83 @@ Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices,
+	for (auto const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Member '.code' does not exist.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Member '.code' is not an array.");
+	for (auto const& codeItem: _json[".code"])
+		solRequire(codeItem.isObject(), AssemblyImportException, "Item of '.code' array is not an object.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (auto const& sourceListItem: _json["sourceList"])
+				solRequire(sourceListItem.isString(), AssemblyImportException, "Item of 'sourceList' array is not of type string.");
+		}
+	} else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Member 'sourceList' is only allowed in root JSON object."
+		);
+
+	shared_ptr<Assembly> result = make_shared<Assembly>(langutil::EVMVersion(), _level == 0, "");
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		for (auto const& it: _json["sourceList"])
+			result->m_sourceList.emplace_back(it.asString());
+	} else
+		result->m_sourceList = _sourceList;
+
+	result->importAssemblyItemsFromJSON(_json[".code"]);
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not of type string.");
+		bytes auxdata{fromHex(_json[".auxdata"].asString())};
+		solRequire(!auxdata.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+		result->m_auxiliaryData = auxdata;
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solRequire(dataIter.key().isString(), AssemblyImportException, "Key inside '.data' is not of type string.");
+			string dataItemID = dataIter.key().asString();
+			Json::Value const& code = data[dataItemID];
+			if (code.isString())
+			{
+				bytes data_value{fromHex(code.asString())};
+				solRequire(
+					!data_value.empty(),
+					AssemblyImportException,
+					"Member '.data' contains a value for '" + dataItemID + "' that is not a valid hexadecimal string.");
+				result->m_data[h256(fromHex(dataItemID))] = fromHex(code.asString());
+			}
+			else if (code.isObject())
+			{
+				shared_ptr<Assembly> subassembly(Assembly::fromJSON(code, result->m_sourceList, _level + 1));
+				solAssert(subassembly);
+				result->m_subs.emplace_back(make_shared<Assembly>(*subassembly));
+				// TODO: this shouldn't be enough for the general case.
+				result->encodeSubPath({0, 0});
```

Comment:
@ekpyron @cameel I just remembered this TODO: right now this is just hardcoded, where at least for all our current test-cases this seem to work fine. But I guess we need to support the general case, right?

Reply:
Or is it possible that some inlining or something like this is happening here? I did not check the resulting yul code yet.

Reply:
If we are not able to have a version that is correct also for the generic case, I have the fear that the import may do something wrong and therefore generate wrong bytecode. I think I will give this another try tomorrow. 

However, another option would be to not do that `result->encodeSubPath({0, 0});` - without that we have four test-cases (`code_access_content.sol`,  `code_access_padding.sol`, `code_access_runtime.sol` and `free_runtimecode.sol`) they will fail with an exception (`solidity/libevmasm/Assembly.cpp(1075): Throw in function vector<size_t> solidity::evmasm::Assembly::decodeSubPath(size_t) const`). We could just ignore them in the import/export script. We could investigate this further after we have received a bug-report. At least without the `encodeSubPath(...);` no wrong bytecode should be generated, the user will just see an exception.

Reply:
I added a command-line test that triggers the current problems with `result->encodeSubPath({0, 0});`. See `test/cmdlineTests/asm_json_export_yul_subobjects`.